### PR TITLE
Introduce RingTokenGenerator configuration in LifecyclerConfig and BasicLifecyclerConfig

### PR DESCRIPTION
**Pull request by duricanikolic**
**What this PR does**:
This PR introduces a `RingTokenGenerator` configuration in `LifecyclerConfig` and `BasicLifecyclerConfig`. `RingTokenGenerator`, if set, specifies the `TokenGenerator` implementation that will be used for token generation. Default value is `nil`, meaning that `RandomTokenGenerator` is used.

**Which issue(s) this PR fixes**:
Part of https://github.com/grafana/mimir/issues/4736

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: ring/lifecycler.go [hunk 1]
```diff
@@ -41,11 +41,12 @@ type LifecyclerConfig struct {
...
-	TokensFilePath           string        `yaml:"tokens_file_path"`
-	Zone                     string        `yaml:"availability_zone"`
-	UnregisterOnShutdown     bool          `yaml:"unregister_on_shutdown" category:"advanced"`
-	ReadinessCheckRingHealth bool          `yaml:"readiness_check_ring_health" category:"advanced"`
+	FinalSleep               time.Duration          `yaml:"final_sleep" category:"advanced"`
+	TokensFilePath           string                 `yaml:"tokens_file_path"`
+	Zone                     string                 `yaml:"availability_zone"`
+	UnregisterOnShutdown     bool                   `yaml:"unregister_on_shutdown" category:"advanced"`
+	ReadinessCheckRingHealth bool                   `yaml:"readiness_check_ring_health" category:"advanced"`
+	SpreadMinimizingZones    flagext.StringSliceCSV `yaml:"spread_minimizing_zones" category:"advanced"`
```
Comment (pracucci) [inline ring/lifecycler.go:49]: This should be `experimental` instead of `advanced`.

File: ring/spread_minimizing_token_generator.go [hunk 2]
```diff
@@ -42,32 +45,36 @@ var (
 )
 
 type SpreadMinimizingConfig struct {
```
Comment (pracucci) [inline ring/spread_minimizing_token_generator.go:47]: Do we need this config at all? I would just pass everything to `NewSpreadMinimizingTokenGenerator()` for now. In practice it doesn't change the number of parameters to pass to `NewSpreadMinimizingTokenGenerator()` given this config struct has only 1 field.

**Review by pracucci — APPROVED** [review]
LGTM!

File: ring/spread_minimizing_token_generator.go (hunk 2 above)
Comment (duricanikolic) [reply ring/spread_minimizing_token_generator.go]: You are right, we don't need this config.
I have removed it, and used a slice of zones instead.

**Pull request merged**